### PR TITLE
add scroll iterator support

### PR DESCRIPTION
**Pull request by 13leaf**
use elasticsearch's scroll endpoint implement iterative large dataset.So we can use like this:

```php
//set limit to null to ensure no limit(default is 10)
foreach(User::find()->limit(NULL)->indexBy('user_id')->iterator() as $uid=>$user){
        //do
}

### CODE REVIEW

File: ScrollIterator.php [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+     * @link http://php.net/manual/en/iterator.rewind.php
+     * @return void Any returned value is ignored.
+     */
+    public function rewind()
+    {
+        if(isset($this->lastScrollId)){
+            $this->command->clearScroll($this->lastScrollId);
+        }
+        if($this->activeQuery->orderBy){
+            $this->scrollOptions['search_type'] = 'scan';
```
Comment (13leaf) [inline ScrollIterator.php]: I'm sorry,I've make a mistake.
We should add search_type=scan when there is no orderBy,to improve performance.
https://www.elastic.co/guide/en/elasticsearch/guide/current/scan-scroll.html

File: ActiveQuery.php [hunk 2]
```diff
@@ -227,6 +227,34 @@ public function one($db = null)
         }
     }
 
+
+    public function createAllModels($result)
+    {
+       if ($this->asArray) {
+          // TODO implement with
```
Comment (samdark) [inline ActiveQuery.php]: ?

File: ActiveQuery.php [hunk 3]
```diff
@@ -227,6 +227,34 @@ public function one($db = null)
...
+          return $result['hits']['hits'];
+       }
+
+       if (empty($result['hits']['hits'])) {
```
Comment (samdark) [inline ActiveQuery.php]: Should it be moved to the very top of the method?

File: ActiveQuery.php [hunk 4]
```diff
@@ -227,6 +227,34 @@ public function one($db = null)
...
+       }
+       foreach ($models as $model) {
+          $model->afterFind();
+       }
+       return $models;
+    }
+
+    public function iterator($db = null,$options = [])
+    {
+       $command=$command = $this->createCommand($db);
```
Comment (samdark) [inline ActiveQuery.php]: Formatting is messed up.

File: ScrollIterator.php [hunk 5]
```diff
@@ -0,0 +1,127 @@
+<?php
+namespace yii\elasticsearch;
+
+/**
+ * User: 13leaf
```
Comment (samdark) [inline ScrollIterator.php]: phpdoc should be corrected.

File: Command.php [hunk 6]
```diff
@@ -319,6 +319,16 @@ public function closeIndex($index)
         return $this->db->post([$index, '_close']);
     }
 
+    public function scroll($scrollId,$options=[])
+    {
+       return $this->db->get(['_search','scroll'],$options,$scrollId);
```
Comment (samdark) [inline Command.php]: Formatting is messed up.

File: ScrollIterator.php (hunk 1 above)
Comment (samdark) [inline ScrollIterator.php]: You can improve the pull request by simple pushing more commits to the same branch.

**Comment by pentium10** [comment]
can we get an official answer here @samdark

**Comment by cebe** [comment]
@pentium10 will review this for the next release

**Comment by cebe** [comment]
duplicate solution as #51

**Comment by cebe** [comment]
closing this as #51 is merged. @13leaf thank you for your contribution!

**Pull request closed without merging**